### DISCUSSION TITLE
[datasource] loki log explore: add streaming log for log query like logcli.

### DISCUSSION THREAD

**Discussion by liguozhong** [grafana/grafana#63829] [Data Sources]
add a input option batch: 50 in log explore ui.
![image](https://user-images.githubusercontent.com/9583245/221745614-1286e250-533b-4e5f-8acb-2f89396005be.png)

log query case, If we want to pull the limit by default: 1000 logs, we can divide it into 20 batches, each with 50 logs. 
In this way, we can continuously output the results in the ui to improve the user experience. 
In other commercial log systems such as new relic, there is a similar "streaming log explore" feature,which can output logs continuously instead of waiting for all logs to arrive.

A similar feature has been implemented in loki's logcli. This is roughly how the go code is written. If grafana/grafana want to do this feature, we need to use type script to write a piece of golang-like code.

The logcli client already uses batches to fetch results.
https://github.com/grafana/loki/blob/main/pkg/logcli/query/query.go#L140

![image](https://user-images.githubusercontent.com/9583245/221745305-f7958f24-1ee7-4db4-b997-7fd3dc4ca5a8.png)

**Comment by liguozhong** [comment]
@gabor

**Comment by gabor** [comment]
hi @liguozhong , there's a feature that we're currently working on, that might be related to this. you need a main-branch grafana, and enable the `lokiQuerySplitting` feature flag. (please note, this is still very much in development)

it will split large logQL queries into single-day chunks, execute them sequentially, so you will see the logs continuously appear.

there's some discussion about it at https://youtu.be/ZFBqNBpLVXY?t=1582 (it works for both metric-data and logs-data)

**Reply by liguozhong** [reply]
> single-day chunks

This is a very good start. It would be even better if `this（single-day chunks）` can be configured. 
In fact, our current query is relatively slow for 3 hours. The expected effect is to query once in `15 minutes`, or once in `5 minutes`.

**Reply by slim-bean** [reply]
The intention is to improve this for sure! this was just a starting point to get the function in place in the frontend and limit complexity for the first iteration.

I think where we would like to end up is an API on Loki that returns the splits to the UI based on the amount of data in the query, the TSDB index can help us make intelligent time splits based on the amount of data matching the label selectors and time range.

**Reply by gabor** [reply]
@liguozhong , we added an experimental config-option now, so the chunk duration can be customized now. it's very very experimental, may change/get-removed, and needs one additional feature flag `lokiQuerySplittingConfig`, but it will do what you wanted to do i think.

**Reply by liguozhong** [reply]
thanks

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). This looks like it got worked on in some capacity, so I am going to close this discussion. Feel free to make an [issue](https://github.com/grafana/grafana/issues) if you cannot find what youre looking for and would like to request the feature.
